File: LEETCODE/EASY/DAY 8/STRING TO INTEGER.py
def myAtoi(s: str) -> int:
    s = s.lstrip()  # remove leading whitespaces
    if not s:
        return 0
    
    # Handle sign
    sign = 1
    if s[0] in ['-', '+']:
        if s[0] == '-':
            sign = -1
        s = s[1:]
    
    # Convert digits
    num = 0
    for char in s:
        if not char.isdigit():
            break
        digit = int(char)
        
        # Handle overflow
        if num > (2**31 - 1) // 10 or (num == (2**31 - 1) // 10 and digit > 7):
            return (2**31 - 1) if sign == 1 else -(2**31)
        
        num = num * 10 + digit
    
    return sign * num

# Example Usage
print(myAtoi("   -42"))  # -42

File: LEETCODE/EASY/DAY 08/STRING TO INTEGER.java
private static final int maxDiv10 = Integer.MAX_VALUE / 10;

public int atoi(String str) {
    int i = 0, n = str.length();
    
    // Skip leading whitespaces
    while (i < n && Character.isWhitespace(str.charAt(i))) i++;
    
    // Handle sign
    int sign = 1;
    if (i < n && str.charAt(i) == '+') {
        i++;
    } else if (i < n && str.charAt(i) == '-') {
        sign = -1;
        i++;
    }
    
    // Convert digits
    int num = 0;
    while (i < n && Character.isDigit(str.charAt(i))) {
        int digit = Character.getNumericValue(str.charAt(i));
        
        // Handle overflow
        if (num > maxDiv10 || (num == maxDiv10 && digit >= 8)) {
            return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;
        }
        
        num = num * 10 + digit;
        i++;
    }
    
    return sign * num;
}
